src/effects/firework_rocket.c

```diff
@@ -1,14 +1,135 @@
 
+void firework_rocket_init(EffectInstance* effect);
+void firework_rocket_render(EffectInstance* effect);
+void firework_rocket_update(EffectInstance* effect);
 void firework_rocket_appendGfx(void* effect);
 
-INCLUDE_ASM(s32, "effects/firework_rocket", firework_rocket_main);
+EffectInstance* firework_rocket_main(s32 arg0, f32 arg1, f32 arg2, f32 arg3, f32 arg4, f32 arg5, f32 arg6, f32 arg7, s32 arg8) {
+    EffectBlueprint bp;
+    EffectInstance* effect;
+    FireworkRocketFXData* data;
+    s32 numParts = 1;
+    s32 i;
 
-void firework_rocket_init(void) {
+    bp.init = firework_rocket_init;
+    bp.update = firework_rocket_update;
+    bp.renderWorld = firework_rocket_render;
+    bp.unk_00 = 0;
+    bp.unk_14 = NULL;
+    bp.effectID = EFFECT_FIREWORK_ROCKET;
+
+    effect = shim_create_effect_instance(&bp);
+    effect->numParts = numParts;
+    data = effect->data.fireworkRocket = shim_general_heap_malloc(numParts * sizeof(*data));
+    ASSERT(effect->data.fireworkRocket != NULL);
+
+    data->unk_00 = arg0;
+    data->unk_20 = 0;
+    if (arg8 <= 0) {
+        data->unk_1C = 1000;
+    } else {
+        data->unk_1C = arg8 + 32;
+    }
+    data->unk_4C = 0;
+    data->unk_30 = 255;
+    data->unk_04 = arg1;
+    data->unk_08 = arg2;
+    data->unk_0C = arg3;
+    data->unk_44 = 0;
+    data->unk_10 = arg4;
+    data->unk_14 = arg5;
+    data->unk_18 = arg6;
+    data->unk_48 = arg7;
+    data->unk_2C = 100;
+    data->unk_38 = 100;
+    data->unk_24 = 255;
+    data->unk_28 = 255;
+    data->unk_34 = 255;
+    data->unk_3C = 150;
+    data->unk_40 = 255;
+
+    for (i = 0; i < 4; i++) {
+        data->unk_50[i] = data->unk_04;
+        data->unk_60[i] = data->unk_08 - 1000.0f;
+        data->unk_70[i] = data->unk_0C;
+        data->unk_80[i] = 0;
+        data->unk_90[i] = 0;
+        data->unk_A0[i] = 0;
+    }
+
+    return effect;
 }
 
-INCLUDE_ASM(s32, "effects/firework_rocket", firework_rocket_update);
+void firework_rocket_init(EffectInstance* effect) {
+}
+
+void firework_rocket_update(EffectInstance* effect) {
+    FireworkRocketFXData* data = effect->data.fireworkRocket;
+    f32 factor;
+    s32 unk_20;
+    s32 i;
+
+    if (effect->flags & 0x10) {
+        effect->flags &= ~0x10;
+        data->unk_1C = 16;
+    }
+
+    if (data->unk_1C < 1000) {
+        data->unk_1C--;
+    }
+
+    data->unk_20++;
+
+    if (data->unk_1C < 0) {
+        shim_remove_effect(effect);
+        return;
+    }
+
+    unk_20 = data->unk_20;
+
+    if (data->unk_1C < 32) {
+        data->unk_30 = data->unk_1C * 8;
+    }
+
+    if (data->unk_4C == 1) {
+        factor = 0.95f;
+        data->unk_04 += data->unk_10;
+        data->unk_08 += data->unk_14;
+        data->unk_0C += data->unk_18;
+        data->unk_10 *= factor;
+        data->unk_14 *= factor;
+        data->unk_18 *= factor;
+        data->unk_44 += (data->unk_48 - data->unk_44) * 0.11;
+        data->unk_14 -= 0.15;
+        return;
+    }
+
+    i = unk_20 & 3;
```

Comment:
Seems like this would be better as `i = unk_20 % 4`

Reply:
I tried it, but it doesn't seem to match.